Radio emitter

pgiri4 opened the pull request:
The radioEmitter branch is now ready to merge with the base develop.  The emitter.py file has been introduced (just like existing askaryan.py)  to include the idea of emitter (as in feature/emitter branch).  And changes in simulation.py file also been updated.  There is also an example inside NuRadioMC/examples/05_pulser_calibration_measurement/radioEmitting_pulser_calibration'

cg-laser commented on NuRadioMC/examples/05_pulser_calibration_measurement/radioEmitting_pulser_calibration/A01generate_pulser_events.py:
```diff
@@ -0,0 +1,77 @@
+    data_sets["emitter_amplitude"] = np.ones(n_events) * 1 * units.V
+
+    # we also have choice for the half width and frequency
+    data_sets["half_of_pulse_width"]= 1.0 * np.ones(n_events) *units.ns        # this will be the width of square and tone_burst signal  
```
can you also prepend an `emitter_` here?

cg-laser commented on NuRadioMC/examples/05_pulser_calibration_measurement/radioEmitting_pulser_calibration/config.yaml:
```diff
@@ -0,0 +1,15 @@
+noise: False  # specify if simulation should be run with or without noise
+sampling_rate: 2.  # sampling rate in GHz used internally in the simulation.
+speedup:
+  minimum_weight_cut: 1.e-5
```
I think you need to deactivate the minimum_weight_cut, i.e. `minimum_weight_cut: 0`. I think for your case it has not effect because the zenith angle is `0`, but better deactivate this feature.

cg-laser commented on NuRadioMC/examples/05_pulser_calibration_measurement/radioEmitting_pulser_calibration/config.yaml:
```diff
@@ -0,0 +1,15 @@
+sampling_rate: 2.  # sampling rate in GHz used internally in the simulation.
+speedup:
+  minimum_weight_cut: 1.e-5
+  delta_C_cut: 100  # 40 degree
```
100rad is no longer 40deg. Adjust the comment.

cg-laser commented on NuRadioMC/examples/05_pulser_calibration_measurement/radioEmitting_pulser_calibration/config.yaml:
```diff
@@ -0,0 +1,15 @@
+  distance_cut: False # Skip ray tracing if vertex distance is greater than a distance cut that is a function of shower energy.
+propagation:
+  ice_model: southpole_2015
+signal:
```
as you don't use an Askaryan model ,better remove this part of the config file to avoid confusions. Same for the `weights` category.

cg-laser commented on NuRadioMC/examples/05_pulser_calibration_measurement/radioEmitting_pulser_calibration/runARA02.py:
```diff
@@ -0,0 +1,91 @@
+        # save the amplitudes to output hdf5 file
+        # save amplitudes per ray tracing solution to hdf5 data output
+        calculateAmplitudePerRaySolution.run(self._evt, self._station, self._det)
+        triggerSimulator.run(evt, station, det,
```
you can give the triggers a name, that might be useful to later better see which one triggered.

cg-laser reviewed (commented):
Thanks for adding this!

I added a few in-line comments. 

The test is still failing, can you have a look? 

I will also go through the simulation.py and try to remove access to the neutrino specific variables, so that the hdf5 does not need to be filled with them.

anelles commented:
Thanks for adding this. It will be a useful module
I would be really useful, if you @pgiri4  used more descriptive commit messages than "updating". In this way, it is very unclear to me whether you addressed any of @cg-laser 's comments. But I feel that @cg-laser still wants to add some work. 
Please check why the tests fail and resolve the conflicts. Otherwise, it makes no sense for me to move to reviewing.

pgiri4 commented:
Hi Anna(@anelles ), sorry for the confusion. I have now resolved the problem regarding the tests fail. But an addition of one more test for radio emitter inside .github/workflows/run_tests.yaml shows a conflict with the develop branch. I hope Christian(@cg-laser ) can suggest me whether this is fine or this should be addressed differently. I have changed the trigger name and removed other unnecessary information from config file as Christian's suggestions.

sjoerd-bouma commented:
> Hi Anna(@anelles ), sorry for the confusion. I have now resolved the problem regarding the tests fail. But an addition of one more test for radio emitter inside .github/workflows/run_tests.yaml shows a conflict with the develop branch. I hope Christian(@cg-laser ) can suggest me whether this is fine or this should be addressed differently. I have changed the trigger name and removed other unnecessary information from config file as Christian's suggestions.

The issue was that another test had been added to develop since you made this PR. I merged it in for you.

sjoerd-bouma commented on NuRadioMC/SignalGen/emitter.py:
```diff
@@ -0,0 +1,153 @@
+        number of samples in the time domain
+    dt: float
+        time bin width, i.e. the inverse of the sampling rate
+    shower_type: string (default "HAD")
+        type of shower, either "HAD" (hadronic), "EM" (electromagnetic) or "TAU" (tau lepton induced)
+        note that TAU showers are currently only implemented in the ARZ2019 model
+    n_index: float
+        index of refraction at interaction vertex
+    R: float
+        distance from vertex to observer
+    model: string
+        specifies the signal model
+        * spherical: a simple signal model of a spherical delta pulse emitter
```
This doesn't seem right. Can you remove the parameters that are not used, and complete the list of available models?

sjoerd-bouma commented on NuRadioMC/SignalGen/emitter.py:
```diff
@@ -0,0 +1,153 @@
+        number of samples in the time domain
+    dt: float
+        time bin width, i.e. the inverse of the sampling rate
+    shower_type: string (default "HAD")
+        type of shower, either "HAD" (hadronic), "EM" (electromagnetic) or "TAU" (tau lepton induced)
+        note that TAU showers are currently only implemented in the ARZ2019 model
+    n_index: float
+        index of refraction at interaction vertex
+    R: float
+        distance from vertex to observer
+    model: string
+        specifies the signal model
+        * spherical: a simple signal model of a spherical delta pulse emitter
```
Same as above

cg-laser commented:
@shallmann I merged this branch with your "save to nur" branch, and made some modifications to not require neutrino specific inputs in the hdf5 file. What do you think? This adds a few conditional `if particle_mode:` statements to the code.

cg-laser commented:
@pgiri4 thanks for adding a working example. Can you make this into a full test, i.e., that you also check that the simulated values agree with a reference? Like we do here https://github.com/nu-radio/NuRadioMC/tree/develop/NuRadioMC/test/SingleEvents You can use the T04validate_almost_equal.py script to compare two hdf5 files of similarity. By adding this you make sure that the example not only runs but keeps generating exactly the same prediction.

cg-laser reviewed (dismissed):
thanks! I think the PR is now ready to get merged. Or is there anything else that needs to be done here?

sjoerd-bouma commented on NuRadioMC/SignalGen/emitter.py:
```diff
@@ -0,0 +1,148 @@
+        time bin width, i.e. the inverse of the sampling rate
+    model: string
+        specifies the signal model
+        * delta_pulse: a simple signal model of a delta pulse emitter
+        * cw : a sinusoidal wave of given frequency
+        * square : a rectangular pulse of given amplituede and width
+        * tone_burst : a short sine wave pulse of given frequency and desired width
+        * idl & hvsp2 : these are the waveforms generated in KU lab and stored in hdf5 files
+        * gaussian : represents a gaussian pulse where sigma is defined through the half width at half maximum
```
Hi, it's very minor, but to prevent errors in generating the documentation (#350 ), could you please add empty lines around the lists in your docstrings?
I.e.
```
       specifies the signal model

        * delta_pulse: a simple signal model of a delta pulse emitter
        * cw : a sinusoidal wave of given frequency
        * square : a rectangular pulse of given amplituede and width
        * tone_burst : a short sine wave pulse of given frequency and desired width
        * idl & hvsp2 : these are the waveforms generated in KU lab and stored in hdf5 files
        * gaussian : represents a gaussian pulse where sigma is defined through the half width at half maximum

      full_output: bool (default False)
```

sjoerd-bouma commented on NuRadioMC/SignalGen/emitter.py:
```diff
@@ -0,0 +1,148 @@
+        time bin width, i.e. the inverse of the sampling rate
+    model: string
+        specifies the signal model
+        * delta_pulse: a simple signal model of a delta pulse emitter
+        * cw : a sinusoidal wave of given frequency
+        * square : a rectangular pulse of given amplituede and width
+        * tone_burst : a short sine wave pulse of given frequency and desired width
+        * idl & hvsp2 : these are the waveforms generated in KU lab and stored in hdf5 files
+        * gaussian : represents a gaussian pulse where sigma is defined through the half width at half maximum
```
See above. Thanks!

sjoerd-bouma reviewed (approved):
Looks good to me now! Seems you got unlucky with the tests - they should complete successfully this time, and then you should be able to merge. Thanks!
